experiments/commands.py

```diff
@@ -0,0 +1,234 @@
+        "the following configs: [`remote_cfg_dir`/example_bc_best.json, "
+        "`remote_cfg_dir`/example_dagger_best.json]"
+        "Why not just supply the pattern '`remote_cfg_dir`/*.json' directly? "
+        "Because the `remote_cfg_dir` directory may not exist on the local machine.",
+    )
+    parser.add_argument(
+        "--seeds",
+        type=str,
+        default="0",
+        help="Comma-delimited list of random seeds to use for each config.",
+    )
+    parser.add_argument(
+        "--output_dir",
+        type=str,
+        required=True,
+        help="Directory in which to store the output training."
+        "If the --remote flag is enabled, "
+        "this directory should be accessible from each container, "
+        "e.g., '/data/output/' if /data is the shared directory.",
+    )
+    parser.add_argument(
+        "--remote",
+        default=False,
+        action="store_true",
+        help="Generate commands to run training scripts "
+        "in containers on the Hofvarpnir cluster.",
+    )
+    # The following flags are only used when the --remote flag is enabled.
+    # Otherwise, they are ignored.
+    parser.add_argument(
+        "--remote_cfg_dir",
+        type=str,
+        default="/data/imitation/benchmarking",
+        help="Path to a directory storing config files "
+        "accessible from each container. ",
+    )
+    parser.add_argument(
+        "--container",
+        type=str,
+        default="hacobe/devbox:imitation",
+        help="The image name to use for the containers.",
+    )
+    args = parser.parse_args()
+
+    cfg_relative_paths = glob.glob(args.cfg_pattern)
+    seeds = [int(s) for s in args.seeds.split(",")]
+    local = not args.remote
+
+    for cfg_relative_path in cfg_relative_paths:
+        cfg_file = os.path.basename(cfg_relative_path)
+        algo_name = _get_algo_name(cfg_file)
+        script_name = _ALGO_NAME_TO_SCRIPT_NAME[algo_name]
+
+        if local:
+            cfg_path = cfg_relative_path
+        else:
+            cfg_path = os.path.join(args.remote_cfg_dir, cfg_file)
+
+        cfg_id = _get_cfg_id(cfg_path)
+
+        for seed in seeds:
+            cmd_id = _CMD_ID_TEMPLATE.format(
+                name=args.name,
+                algo_name=algo_name,
+                seed=seed,
+                cfg_id=cfg_id,
+            )
+
+            file_storage = os.path.join(args.output_dir, "sacred", cmd_id)
+
+            train_cmd = _TRAIN_CMD_TEMPLATE.format(
+                script_name=script_name,
+                algo_name=algo_name,
+                name=args.name,
+                cfg_path=cfg_path,
+                seed=seed,
+                file_storage=file_storage,
+                log_root=args.output_dir,
+            )
+
+            if local:
+                print(train_cmd)
+                continue
+
+            # Escape spaces.
+            command = train_cmd.replace(" ", "\\ ")
```

Comment:
This makes commands output quite hard to read, and might still fail in presence of some other special characters. How about changing format to have '' around parameters, so `--name '{name}'` etc?

Reply:
I tried out one command escaping spaces and one command enclosed in quotes.

(1) `ctl job run --command echo\ aaa --container hacobe/devbox:imitation --login --high-priority --force-pull --never-restart --name echo-test`

(2) `ctl job run --command "echo 'aaa'" --container hacobe/devbox:imitation --login --high-priority --force-pull --never-restart --name echo-test2`

(1) completes quickly and (2) gets stuck in Pending.

Can we leave the escaping spaces as is?

Reply:
I agree that the escaped spaces make the command less readable. I pushed a [commit](https://github.com/HumanCompatibleAI/imitation/pull/654/commits/b0e133ccbf85c309a26577453c50d37250d032f2) that escapes double quotes in the training command and encloses it within the cluster command in double quotes.

I don't want to enclose other flag values in single quotes, because then an environment variable like $USER in the flag value won't get evaluated. I also don't want to enclose other flag values in double quotes, because that again clutters the command. In the commit message, I show an example of how you would run commands.py if you wanted to use a special character like a space in the output directory.

Let me know if this approach works for you.

Reply:
To check I understand: the point is most other flags won't have spaces in them, so enclosing them all in double quotes would just clutter things? That seems like a reasonable call. I worry a bit that this is fragile (e.g. what if IFS was set to things other than just spaces), but for this application readability may trump robustness.

Reply:
Yeah that's my thinking. And a user can always enclose the flag values they pass in double quotes like this --output-dir="\"/path/with/special characters\"" if they want the flag in the generated command to be enclosed in double quotes like this --logging_dir.log_root="/path/with/special characters".